[Improvement][UT] Upgrade jUnit to 5.+ (#10976)

EricGao888 opened the pull request:
## Purpose of the pull request

* Update jUnit to 5.+ following the official jUnit migration instructions: https://github.com/junit-team/junit5-samples/blob/main/junit5-migration-maven/pom.xml
* Refactor `AlertServerTest` with jUnit 5 as an example.
* This PR closes: #10976 

## Brief change log

* Already described above.

## Verify this pull request

* Verified by passing all UTs.

EricGao888 committed:
[Improvement][UT] Upgrade jUnit to 5.+ (#10976)

EricGao888 committed:
Refactor AlertServerTest with jUnit-5 as an example

EricGao888 committed:
Resolve dependency conflicts

EricGao888 committed:
Remove redundant commented code

kezhenxu94 commented on dolphinscheduler-api/pom.xml:
```diff
@@ -119,86 +130,80 @@
             <artifactId>hibernate-validator</artifactId>
         </dependency>
-
+        
```
Can you remove the indentations? And add this to spotless?

EricGao888 replied:
Sure, a bit wired though, 
was `indent_empty_lines` not working? https://github.com/apache/dolphinscheduler/blob/983f9b41bf86bb07a3b439eadf4e7bc201d4def7/style/spotless_dolphinscheduler_formatter.xml#L26

EricGao888 replied:
Nailed it, it was overwritten here: https://github.com/apache/dolphinscheduler/blob/983f9b41bf86bb07a3b439eadf4e7bc201d4def7/pom.xml#L657

EricGao888 committed:
Fix code style

caishunfeng commented on pom.xml:
```diff
@@ -309,23 +309,57 @@
+            <dependency>
+                <groupId>org.junit</groupId>
+                <artifactId>junit-bom</artifactId>
+                <version>5.9.0</version>
```
Is it better to set version in properties for unified version management?

EricGao888 replied:
Yes, I will add it to properties. Same as that of `mockito`.

caishunfeng commented on dolphinscheduler-api/src/test/java/org/apache/dolphinscheduler/api/controller/TaskInstanceControllerTest.java:
```diff
@@ -42,15 +42,19 @@
 import org.springframework.util.LinkedMultiValueMap;
 import org.springframework.util.MultiValueMap;
 
+//@RunWith(MockitoJUnitRunner.class)
```
Should this comment be recover?

EricGao888 replied:
Sure, I will recover it.

caishunfeng commented:
> SonarCloud Quality Gate failed.    [![Quality Gate failed](https://camo.githubusercontent.com/4ea51c1f64ee3746f631653a02ab678ca6a3efb5f5cb474402faed2e3dcf90b5/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f5175616c6974794761746542616467652f6661696c65642d313670782e706e67)](https://sonarcloud.io/dashboard?id=apache-dolphinscheduler&pullRequest=11332)
> 
> [![Bug](https://camo.githubusercontent.com/4c6102327f5a954f9c8acaf2e2714183157a9e41717b371b2cd585cf25057310/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f6275672d313670782e706e67)](https://sonarcloud.io/project/issues?id=apache-dolphinscheduler&pullRequest=11332&resolved=false&types=BUG) [![C](https://camo.githubusercontent.com/57cf2cd81158730ed1dd8be1e93a41a2feb8b35dd75e5601402ba5f961f6ec18/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f526174696e6742616467652f432d313670782e706e67)](https://sonarcloud.io/project/issues?id=apache-dolphinscheduler&pullRequest=11332&resolved=false&types=BUG) [21 Bugs](https://sonarcloud.io/project/issues?id=apache-dolphinscheduler&pullRequest=11332&resolved=false&types=BUG) [![Vulnerability](https://camo.githubusercontent.com/3ba1ee49636ffc3427e38649a9f8a65ee392f28e8a662fcf96ce24cefbb520e9/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f76756c6e65726162696c6974792d313670782e706e67)](https://sonarcloud.io/project/issues?id=apache-dolphinscheduler&pullRequest=11332&resolved=false&types=VULNERABILITY) [![E](https://camo.githubusercontent.com/ca3e5c9e7ad5fd04244d2d793976efbe479a024b145a815384556548a9884b5f/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f526174696e6742616467652f452d313670782e706e67)](https://sonarcloud.io/project/issues?id=apache-dolphinscheduler&pullRequest=11332&resolved=false&types=VULNERABILITY) [9 Vulnerabilities](https://sonarcloud.io/project/issues?id=apache-dolphinscheduler&pullRequest=11332&resolved=false&types=VULNERABILITY) [![Security Hotspot](https://camo.githubusercontent.com/fb735cbe76f8d5e1679c76ce83b740ceb1eaf62de4f7bf88623dc9953261aff7/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f73656375726974795f686f7473706f742d313670782e706e67)](https://sonarcloud.io/project/security_hotspots?id=apache-dolphinscheduler&pullRequest=11332&resolved=false&types=SECURITY_HOTSPOT) [![E](https://camo.githubusercontent.com/ca3e5c9e7ad5fd04244d2d793976efbe479a024b145a815384556548a9884b5f/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f526174696e6742616467652f452d313670782e706e67)](https://sonarcloud.io/project/security_hotspots?id=apache-dolphinscheduler&pullRequest=11332&resolved=false&types=SECURITY_HOTSPOT) [37 Security Hotspots](https://sonarcloud.io/project/security_hotspots?id=apache-dolphinscheduler&pullRequest=11332&resolved=false&types=SECURITY_HOTSPOT) [![Code Smell](https://camo.githubusercontent.com/8fe18b2dfb6f7d4e44582f281b29f617eb5ae07c248d2002ca586e91da219212/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f636f64655f736d656c6c2d313670782e706e67)](https://sonarcloud.io/project/issues?id=apache-dolphinscheduler&pullRequest=11332&resolved=false&types=CODE_SMELL) [![A](https://camo.githubusercontent.com/1cba125a897d7fa47033a3b3b2be2bbee680d34d4f004a215564659b853fb201/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f526174696e6742616467652f412d313670782e706e67)](https://sonarcloud.io/project/issues?id=apache-dolphinscheduler&pullRequest=11332&resolved=false&types=CODE_SMELL) [2101 Code Smells](https://sonarcloud.io/project/issues?id=apache-dolphinscheduler&pullRequest=11332&resolved=false&types=CODE_SMELL)
> 
> [![38.0%](https://camo.githubusercontent.com/8f6686cb3d0947decfd92115e15fd4cd1d205e261a5e90a28f193db2a2ad4719/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f436f76657261676543686172742f32352d313670782e706e67)](https://sonarcloud.io/component_measures?id=apache-dolphinscheduler&pullRequest=11332&metric=new_coverage&view=list) [38.0% Coverage](https://sonarcloud.io/component_measures?id=apache-dolphinscheduler&pullRequest=11332&metric=new_coverage&view=list) [![2.4%](https://camo.githubusercontent.com/8047c63e1f9ed03f63001e1eadce4676bade3e0f83ec690a9c625287796248a6/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f4475706c69636174696f6e732f332d313670782e706e67)](https://sonarcloud.io/component_measures?id=apache-dolphinscheduler&pullRequest=11332&metric=new_duplicated_lines_density&view=list) [2.4% Duplication](https://sonarcloud.io/component_measures?id=apache-dolphinscheduler&pullRequest=11332&metric=new_duplicated_lines_density&view=list)

Please check the sonar result

EricGao888 commented:
BTW, there are two ways to fully migrate those tests from jUnit 4 to jUnit 5:

1. We keep both versions of jUnit dependencies and migrate those UTs gradually with `DSIP-10` #10573.  
2. Remove the dependency of jUnit 4 and migrate all the UTs with IDE `inspections` in this PR.

However, I think the second method is a bit risky. 

WDYT? @caishunfeng @kezhenxu94 @SbloodyS @ruanwenjun

ruanwenjun commented:
> BTW, there are two ways to fully migrate those tests from jUnit 4 to jUnit 5:
> 
> 1. We keep both versions of jUnit dependencies and migrate those UTs gradually with `DSIP-10` [[DSIP-10][Unit Tests] Improve DolphinScheduler unit tests #10573](https://github.com/apache/dolphinscheduler/issues/10573).
> 2. Remove the dependency of jUnit 4 and migrate all the UTs with IDE `inspections` in this PR.
> 
> However, I think the second method is a bit risky.
> 
> WDYT? @caishunfeng @kezhenxu94 @SbloodyS @ruanwenjun

I prefer the second way.

EricGao888 commented:
> > BTW, there are two ways to fully migrate those tests from jUnit 4 to jUnit 5:
> > 
> > 1. We keep both versions of jUnit dependencies and migrate those UTs gradually with `DSIP-10` [[DSIP-10][Unit Tests] Improve DolphinScheduler unit tests #10573](https://github.com/apache/dolphinscheduler/issues/10573).
> > 2. Remove the dependency of jUnit 4 and migrate all the UTs with IDE `inspections` in this PR.
> > 
> > However, I think the second method is a bit risky.
> > WDYT? @caishunfeng @kezhenxu94 @SbloodyS @ruanwenjun
> 
> I prefer the second way.

I just tried migrating all the UTs and got blocked.

Specifically speaking, there are two risks:

1. We need to switch all the UTs to use jUnit 5 and `Spotless` will automatically reformat them. I've tried it, there would be 333 files modified. 
2. We used `powermock` quite a lot in the project and `powermock` does not support `jUnit 5` as stated in https://github.com/apache/dolphinscheduler/issues/10976#issuecomment-1207427898. It is hardly possible to refactor all the code using `powermock` in one PR.

The real issue blocking me is the second one. May I ask whether there is a good way, or some kind of workaround for the second point? @ruanwenjun @kezhenxu94

ruanwenjun commented:
> > > BTW, there are two ways to fully migrate those tests from jUnit 4 to jUnit 5:
> > > 
> > > 1. We keep both versions of jUnit dependencies and migrate those UTs gradually with `DSIP-10` [[DSIP-10][Unit Tests] Improve DolphinScheduler unit tests #10573](https://github.com/apache/dolphinscheduler/issues/10573).
> > > 2. Remove the dependency of jUnit 4 and migrate all the UTs with IDE `inspections` in this PR.
> > > 
> > > However, I think the second method is a bit risky.
> > > WDYT? @caishunfeng @kezhenxu94 @SbloodyS @ruanwenjun
> > 
> > 
> > I prefer the second way.
> 
> I just tried migrating all the UTs and got blocked.
> 
> Specifically speaking, there are two risks:
> 
> 1. We need to switch all the UTs to use jUnit 5 and `Spotless` will automatically reformat them. I've tried it, there would be 333 files modified.
> 2. We used `powermock` quite a lot in the project and `powermock` does not support `jUnit 5` as stated in [[Improvement][UT] Upgrade junit to 5.+ #10976 (comment)](https://github.com/apache/dolphinscheduler/issues/10976#issuecomment-1207427898). It is hardly possible to refactor all the code using `powermock` in one PR.
> 
> The real issue blocking me is the second one. May I ask whether there is a good way, or some kind of workaround for the second point? @ruanwenjun @kezhenxu94

I just worry if we still keep the Junit4, we need to constantly remind contributors to use Junit5.

kezhenxu94 commented:
> > > BTW, there are two ways to fully migrate those tests from jUnit 4 to jUnit 5:
> > > 
> > > 1. We keep both versions of jUnit dependencies and migrate those UTs gradually with `DSIP-10` [[DSIP-10][Unit Tests] Improve DolphinScheduler unit tests #10573](https://github.com/apache/dolphinscheduler/issues/10573).
> > > 2. Remove the dependency of jUnit 4 and migrate all the UTs with IDE `inspections` in this PR.
> > > 
> > > However, I think the second method is a bit risky.
> > > WDYT? @caishunfeng @kezhenxu94 @SbloodyS @ruanwenjun
> > 
> > 
> > I prefer the second way.
> 
> I just tried migrating all the UTs and got blocked.
> 
> Specifically speaking, there are two risks:
> 
> 1. We need to switch all the UTs to use jUnit 5 and `Spotless` will automatically reformat them. I've tried it, there would be 333 files modified.

This is totally OK to me, "migrations" are always huge changes...

> 2. We used `powermock` quite a lot in the project and `powermock` does not support `jUnit 5` as stated in [[Improvement][UT] Upgrade junit to 5.+ #10976 (comment)](https://github.com/apache/dolphinscheduler/issues/10976#issuecomment-1207427898). It is hardly possible to refactor all the code using `powermock` in one PR.
> 
> The real issue blocking me is the second one. May I ask whether there is a good way, or some kind of workaround for the second point? @ruanwenjun @kezhenxu94

I hope we could finally get rid of powermock one day, did you find any alternative to Powermock in JUnit5 platform? Or it's just impossible?

kezhenxu94 commented:
> I just worry if we still keep the Junit4, we need to constantly remind contributors to use Junit5.


Agree, same concern to me

EricGao888 committed:
Remove jUnit4 dependency

EricGao888 committed:
Fix transitive dependency bug

EricGao888 commented:
@ruanwenjun @kezhenxu94 @caishunfeng 
Finally I managed to remove the dependency of `jUnit 4` since it is wired, ugly and confusing to have two versions of `jUnit` in `pom.xml`.

I have the same worries that ` if we still keep the Junit4, we need to constantly remind contributors to use Junit5`.

However, even though I have removed the dependency of `jUnit 4`, contributors still could write `jUnit 4` fashion UTs as long as we use `junit-vintage-engine`. It is quite difficult to migrate all legacy tests in this PR. I will keep working on this but I prefer doing it iteratively. 

About `Powermock`, It seems we could use high version (greater than 3.4.0) `Mockito` to mock static methods and I've already added the dependency in this PR. I need more time to find out whether the latest `Mockito` could fully replace `Powermock` since we use `Powermock` quite a lot previously. Also, I find someone use `bytebuddy` to do some trick but I'm not sure whether this trick could apply here: https://git.stklcode.de/stklcode/juraclient/commit/c9444bbc971cd36f80824c3e7ec9baf8ecfbd676. 
But I believe we could get rid of `Powermock` as long as we set our minds to : ) 
https://wttech.blog/blog/2020/mocking-static-methods-made-possible-in-mockito-3.4.0/

For the rest of the legacy tests, we could migrate some of them easily with `Intellij`, but still, there are some complicated ones we need to do it manually. 

Above all, there are two action items after this PR, if you agree:

- [ ] Remove dependency of `Powermock`. https://github.com/apache/dolphinscheduler/issues/11405.
- [ ] Migrate legacy code and replace `junit-vintage-engine` with `junit-jupiter-engine`.

Thanks

EricGao888 commented:
FYI here is a list of UT files which uses `Powermock`:

```txt
Found Occurrences in Directory dolphinscheduler with mask '*.java'  (192 usages found)
    Unclassified  (192 usages found)
        dolphinscheduler-alert-server  (1 usage found)
            org.apache.dolphinscheduler.alert  (1 usage found)
                AlertServerTest.java  (1 usage found)
        dolphinscheduler-api  (27 usages found)
            org.apache.dolphinscheduler.api.controller  (5 usages found)
                DataAnalysisControllerTest.java  (1 usage found)
                ProcessInstanceControllerTest.java  (1 usage found)
                ProcessTaskRelationControllerTest.java  (1 usage found)
                ResourcesControllerTest.java  (1 usage found)
                WorkerGroupControllerTest.java  (1 usage found)
            org.apache.dolphinscheduler.api.interceptor  (2 usages found)
                RateLimitInterceptorTest.java  (2 usages found)
            org.apache.dolphinscheduler.api.service  (20 usages found)
                BaseServiceTest.java  (3 usages found)
                DataAnalysisServiceTest.java  (2 usages found)
                DataSourceServiceTest.java  (4 usages found)
                LoggerServiceTest.java  (1 usage found)
                ResourcesServiceTest.java  (4 usages found)
                SchedulerServiceTest.java  (2 usages found)
                TenantServiceTest.java  (1 usage found)
                UdfFuncServiceTest.java  (3 usages found)
        dolphinscheduler-common  (10 usages found)
            org.apache.dolphinscheduler.common.utils  (10 usages found)
                CommonUtilsTest.java  (2 usages found)
                FileUtilsTest.java  (3 usages found)
                HadoopUtilsTest.java  (4 usages found)
                NetUtilsTest.java  (1 usage found)
        dolphinscheduler-datasource-api  (9 usages found)
            org.apache.dolphinscheduler.plugin.datasource.api.client  (4 usages found)
                CommonDataSourceClientTest.java  (4 usages found)
            org.apache.dolphinscheduler.plugin.datasource.api.utils  (5 usages found)
                CommonUtilsTest.java  (5 usages found)
        dolphinscheduler-datasource-clickhouse  (4 usages found)
            org.apache.dolphinscheduler.plugin.datasource.clickhouse  (4 usages found)
                ClickHouseDataSourceChannelTest.java  (4 usages found)
        dolphinscheduler-datasource-db2  (7 usages found)
            org.apache.dolphinscheduler.plugin.datasource.db2  (4 usages found)
                DB2DataSourceChannelTest.java  (4 usages found)
            org.apache.dolphinscheduler.plugin.datasource.db2.param  (3 usages found)
                Db2DataSourceProcessorTest.java  (3 usages found)
        dolphinscheduler-datasource-hive  (7 usages found)
            org.apache.dolphinscheduler.plugin.datasource.hive  (4 usages found)
                HiveDataSourceChannelTest.java  (4 usages found)
            org.apache.dolphinscheduler.plugin.datasource.hive.param  (3 usages found)
                HiveDataSourceProcessorTest.java  (3 usages found)
        dolphinscheduler-datasource-mysql  (14 usages found)
            org.apache.dolphinscheduler.plugin.datasource.mysql  (4 usages found)
                MySQLDataSourceChannelTest.java  (4 usages found)
            org.apache.dolphinscheduler.plugin.datasource.mysql.param  (3 usages found)
                MySQLDataSourceProcessorTest.java  (3 usages found)
            org.apache.dolphinscheduler.plugin.datasource.mysql.provider  (3 usages found)
                JDBCDataSourceProviderTest.java  (3 usages found)
            org.apache.dolphinscheduler.plugin.datasource.mysql.utils  (4 usages found)
                DataSourceUtilsTest.java  (4 usages found)
        dolphinscheduler-datasource-oracle  (7 usages found)
            org.apache.dolphinscheduler.plugin.datasource.oracle  (4 usages found)
                OracleDataSourceChannelTest.java  (4 usages found)
            org.apache.dolphinscheduler.plugin.datasource.oracle.param  (3 usages found)
                OracleDataSourceProcessorTest.java  (3 usages found)
        dolphinscheduler-datasource-postgresql  (7 usages found)
            org.apache.dolphinscheduler.plugin.datasource.postgresql  (4 usages found)
                PostgreSQLDataSourceChannelTest.java  (4 usages found)
            org.apache.dolphinscheduler.plugin.datasource.postgresql.param  (3 usages found)
                PostgreSQLDataSourceProcessorTest.java  (3 usages found)
        dolphinscheduler-datasource-presto  (7 usages found)
            org.apache.dolphinscheduler.plugin.datasource.presto  (4 usages found)
                PrestoDataSourceChannelTest.java  (4 usages found)
            org.apache.dolphinscheduler.plugin.datasource.presto.param  (3 usages found)
                PrestoDataSourceProcessorTest.java  (3 usages found)
        dolphinscheduler-datasource-redshift  (3 usages found)
            org.apache.dolphinscheduler.plugin.datasource.redshift.param  (3 usages found)
                RedshiftDataSourceProcessorTest.java  (3 usages found)
        dolphinscheduler-datasource-spark  (7 usages found)
            org.apache.dolphinscheduler.plugin.datasource.spark  (4 usages found)
                SparkDataSourceChannelTest.java  (4 usages found)
            org.apache.dolphinscheduler.plugin.datasource.spark.param  (3 usages found)
                SparkDataSourceProcessorTest.java  (3 usages found)
        dolphinscheduler-datasource-sqlserver  (7 usages found)
            org.apache.dolphinscheduler.plugin.datasource.sqlserver  (4 usages found)
                SQLServerDataSourceChannelTest.java  (4 usages found)
            org.apache.dolphinscheduler.plugin.datasource.sqlserver.param  (3 usages found)
                SQLServerDataSourceProcessorTest.java  (3 usages found)
        dolphinscheduler-log-server  (3 usages found)
            org.apache.dolphinscheduler.server.log  (3 usages found)
                LoggerRequestProcessorTest.java  (3 usages found)
        dolphinscheduler-master  (18 usages found)
            org.apache.dolphinscheduler.server.master  (3 usages found)
                SubProcessTaskTest.java  (3 usages found)
            org.apache.dolphinscheduler.server.master.processor  (5 usages found)
                CacheProcessorTest.java  (1 usage found)
                TaskAckProcessorTest.java  (3 usages found)
                TaskKillResponseProcessorTest.java  (1 usage found)
            org.apache.dolphinscheduler.server.master.registry  (3 usages found)
                MasterRegistryClientTest.java  (3 usages found)
            org.apache.dolphinscheduler.server.master.runner  (4 usages found)
                MasterTaskExecThreadTest.java  (1 usage found)
                WorkflowExecuteRunnableTest.java  (3 usages found)
            org.apache.dolphinscheduler.server.master.service  (3 usages found)
                FailoverServiceTest.java  (3 usages found)
        dolphinscheduler-server  (4 usages found)
            org.apache.dolphinscheduler.server.utils  (4 usages found)
                ProcessUtilsTest.java  (4 usages found)
        dolphinscheduler-service  (8 usages found)
            org.apache.dolphinscheduler.service.alert  (4 usages found)
                AlertClientServiceTest.java  (3 usages found)
                ProcessAlertManagerTest.java  (1 usage found)
            org.apache.dolphinscheduler.service.log  (3 usages found)
                LogClientServiceTest.java  (3 usages found)
            org.apache.dolphinscheduler.service.process  (1 usage found)
                ProcessServiceTest.java  (1 usage found)
        dolphinscheduler-task-dvc  (5 usages found)
            org.apache.dolphinscheduler.plugin.task.dvc  (5 usages found)
                DvcTaskTest.java  (5 usages found)
        dolphinscheduler-task-emr  (8 usages found)
            org.apache.dolphinscheduler.plugin.task.emr  (8 usages found)
                EmrAddStepsTaskTest.java  (4 usages found)
                EmrJobFlowTaskTest.java  (4 usages found)
        dolphinscheduler-task-jupyter  (5 usages found)
            org.apache.dolphinscheduler.plugin.task.jupyter  (5 usages found)
                JupyterTaskTest.java  (5 usages found)
        dolphinscheduler-task-mlflow  (5 usages found)
            org.apache.dolphinler.plugin.task.mlflow  (5 usages found)
                MlflowTaskTest.java  (5 usages found)
        dolphinscheduler-task-openmldb  (1 usage found)
            org.apache.dolphinscheduler.plugin.task.openmldb  (1 usage found)
                OpenmldbTaskTest.java  (1 usage found)
        dolphinscheduler-task-python  (1 usage found)
            org.apache.dolphinscheduler.plugin.task.python  (1 usage found)
                PythonTaskTest.java  (1 usage found)
        dolphinscheduler-task-sagemaker  (5 usages found)
            org.apache.dolphinscheduler.plugin.task.sagemaker  (5 usages found)
                SagemakerTaskTest.java  (5 usages found)
        dolphinscheduler-task-spark  (4 usages found)
            org.apache.dolphinscheduler.plugin.task.spark  (4 usages found)
                SparkTaskTest.java  (4 usages found)
        dolphinscheduler-task-zeppelin  (4 usages found)
            org.apache.dolphinscheduler.plugin.task.zeppelin  (4 usages found)
                ZeppelinTaskTest.java  (4 usages found)
        dolphinscheduler-worker  (4 usages found)
            org.apache.dolphinscheduler.server.worker.processor  (3 usages found)
                TaskDispatchProcessorTest.java  (3 usages found)
            org.apache.dolphinscheduler.server.worker.runner  (1 usage found)
                TaskExecuteThreadTest.java  (1 usage found)

```

EricGao888 committed:
[Feature][Style] Enable spotless to manage imports (#11458)

EricGao888 commented:
I investigated the migration from `jUnit4` to `jUnit5` in other open-source projects, such as `Apache Flink`. `Apache Flink` community once tried to migrate all stuff from 4 to 5 once for all but finally gave up and decided to do it gradually. Currently, as shown in 
https://github.com/apache/flink/blob/1232629c80cbb64eb4ca9f6c95d6c5c1a2e8e82d/pom.xml#L147-L148 , they have both versions in the project.

FYI, here is the mailing thread where they discussed about the migration: https://www.mail-archive.com/dev@flink.apache.org/msg47657.html

WDYT @kezhenxu94 @ruanwenjun @caishunfeng @SbloodyS

EricGao888 commented:
> I investigated the migration from `jUnit4` to `jUnit5` in other open-source projects, such as `Apache Flink`. `Apache Flink` community once tried to migrate all stuff from 4 to 5 once for all but finally gave up and decided to do it gradually. Currently, as shown in https://github.com/apache/flink/blob/1232629c80cbb64eb4ca9f6c95d6c5c1a2e8e82d/pom.xml#L147-L148 , they have both versions in the project.
> 
> FYI, here is the mailing thread where they discussed about the migration: https://www.mail-archive.com/dev@flink.apache.org/msg47657.html
> 
> WDYT @kezhenxu94 @ruanwenjun @caishunfeng @SbloodyS

cc @zhongjiajie

zhongjiajie commented:
> > I investigated the migration from `jUnit4` to `jUnit5` in other open-source projects, such as `Apache Flink`. `Apache Flink` community once tried to migrate all stuff from 4 to 5 once for all but finally gave up and decided to do it gradually. Currently, as shown in https://github.com/apache/flink/blob/1232629c80cbb64eb4ca9f6c95d6c5c1a2e8e82d/pom.xml#L147-L148 , they have both versions in the project.
> > FYI, here is the mailing thread where they discussed about the migration: https://www.mail-archive.com/dev@flink.apache.org/msg47657.html
> > WDYT @kezhenxu94 @ruanwenjun @caishunfeng @SbloodyS
> 
> cc @zhongjiajie

Maybe there are other historical reasons of flink,dolphischeduler do not have many unitest currently, IMO if we can migrate all of them to junit5 now it is the time

EricGao888 commented:
> > > I investigated the migration from `jUnit4` to `jUnit5` in other open-source projects, such as `Apache Flink`. `Apache Flink` community once tried to migrate all stuff from 4 to 5 once for all but finally gave up and decided to do it gradually. Currently, as shown in https://github.com/apache/flink/blob/1232629c80cbb64eb4ca9f6c95d6c5c1a2e8e82d/pom.xml#L147-L148 , they have both versions in the project.
> > > FYI, here is the mailing thread where they discussed about the migration: https://www.mail-archive.com/dev@flink.apache.org/msg47657.html
> > > WDYT @kezhenxu94 @ruanwenjun @caishunfeng @SbloodyS
> > 
> > 
> > cc @zhongjiajie
> 
> Maybe there are other historical reasons of flink,dolphischeduler do not have many unitest currently, IMO if we can migrate all of them to junit5 now it is the time

@zhongjiajie The workload to migrate all of them is quite large. Currently, because of those `powermock`, I could not give an exact estimate of the specific workload. If we decide to do so, first of all, we need to remove all the code related to powermock #11405 . Then, we need to get back to this PR to add dependency on jUnit5 and manually migrate the complex part which IDE could not do automatically e.g. change `expected = Exception.class` to `Assertions.assertThrows`. After these two done, we could try to use `Intellij` to migrate the rest and see how many of the rest will break and decide what to do next.

EricGao888 commented:
> > I just worry if we still keep the Junit4, we need to constantly remind contributors to use Junit5.
> 
> Agree, same concern to me

@ruanwenjun @kezhenxu94 BTW, for the concern that contributors may write UTs with `jUnit 4`, we could add `regex` in `Spotless` check rules to prevent them from using `jUnit 4`. This will work in CI too. The same with `Powermock`.

kezhenxu94 commented:
Powermock is getting more and more annoying such as https://github.com/apache/dolphinscheduler/runs/7929368593?check_suite_focus=true 

https://github.com/powermock/powermock/issues/1112

I'd rather just review and remove all those tests involving Powermock...

EricGao888 commented:
> Powermock is getting more and more annoying such as https://github.com/apache/dolphinscheduler/runs/7929368593?check_suite_focus=true
> 
> [powermock/powermock#1112](https://github.com/powermock/powermock/issues/1112)
> 
> I'd rather just review and remove all those tests involving Powermock...

@kezhenxu94 To follow up on this comment, I did some experiments on replacing `Powermock` with `Mockito` and it worked well, see #11588. I think it is possible to remove the dependency of `Powermock`. 

For some more difficult cases such as:
1. Legacy code using some external APIs which are really hard to mock
2. Legacy code instantiating stuff with `new` keyword 

We could:
1. Encapsulate the external APIs 
2. Inject dependencies as below: 
![image](https://user-images.githubusercontent.com/34905992/185851674-37204ca1-9eb6-4b56-9631-abb41aac4b2f.png)
 
 https://stackoverflow.com/questions/15052984/what-is-the-difference-between-mocking-and-spying-when-using-mockito#:~:text=about%20partial%20mocks.-,Mockito.,passed%20to%20spy()%20method.

kezhenxu94 commented:
Good to hear we can remove powermock 🎉

ruanwenjun reviewed (approved):
LGTM, in this PR we upgrade the unit version to `5.9.0`, and we still keep the powermock(Hope we can remove powermock in the later pr).

EricGao888 commented:
> LGTM, in this PR we upgrade the unit version to `5.9.0`, and we still keep the powermock(Hope we can remove powermock in the later pr).

@ruanwenjun Certainly we will remove `powermock` in the whole project soon and I will add a spotless step to block it in DS. Removing `powermock` is of the top priority for DSIP-10 #10573 .  BTW, I just found that there had been both jUnit 4 and jUnit 5 style tests in DS project for a long time. see: https://github.com/apache/dolphinscheduler/pull/12062#issuecomment-1251952599

SbloodyS reviewed (approved):
+1
